File: leetcode/hot100-java/group_2/code23/reverse_list.java
/*
 * 206. 反转链表
 *
 * 给定单链表的头节点 head，原地反转链表中所有 next 指针，并返回反转后的新
 * 头节点。空链表应返回 null。
 *
 * 算法实现说明：
 * 1. previous 指向已经反转好的链表头，current 指向下一个等待处理的节点。
 * 2. 每轮先用 following 保存 current.next，再把 current.next 改指向 previous，防止丢失后续链表。
 * 3. 随后同步前移 previous 和 current；current 变为 null 时，previous 就是完整反转后的新头节点。
 *
 * 时间复杂度：O(n)；空间复杂度：O(1)。
 */
class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode previous = null;
        ListNode current = head;
        while (current != null) {
            ListNode following = current.next;
            current.next = previous;
            previous = current;
            current = following;
        }
        return previous;
    }
}

File: codeTop/go/groupI/reverseList.go
package groupI

/*
题目：206. 反转链表
链接：https://leetcode.cn/problems/reverse-linked-list/

题目内容：
给定单链表的头节点 head，反转链表中所有节点的连接方向，并返回反转后的新头节点。
空链表应返回 nil。
*/

import "algo/structure"

type ListNode = structure.ListNode

func reverseList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	var pre *ListNode
	cur := head
	for cur != nil {
		next := cur.Next
		cur.Next = pre
		pre = cur
		cur = next
	}
	return pre
}
